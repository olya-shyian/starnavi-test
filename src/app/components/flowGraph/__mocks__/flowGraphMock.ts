import { FormattedFilms } from "@/app/interfaces/IFilms";

export const heroeInfoMock: FormattedFilms[] = [
  {
    title: "A New Hope",
    id: "1",
    starships: [
      { id: "1", name: "X-Wing Starfighter" },
      { id: "2", name: "Millennium Falcon" },
    ],
  },
  {
    title: "The Empire Strikes Back",
    id: "2",
    starships: [{ id: "3", name: "Snowspeeder" }],
  },
];

export const heroeInfoWithNoStarshipsMock: FormattedFilms[] = [
  {
    title: "A New Hope",
    id: "1",
  },
  {
    title: "The Empire Strikes Back",
    id: "2",
  },
];

export const starshipsMock = [
  {
    MGLT: "75",
    cargo_capacity: "100000",
    consumables: "2 months",
    cost_in_credits: "3500000",
    created: "2014-12-12T11:05:37.420000Z",
    crew: "4",
    edited: "2014-12-22T17:35:44.431407Z",
    hyperdrive_rating: "1.0",
    length: "34.37",
    manufacturer: "Kuat Drive Yards",
    max_atmosphering_speed: "950",
    model: "Sentinel-class landing craft",
    name: "Sentinel-class landing craft",
    passengers: "75",
    films: [1, 2], // Example film IDs
    pilots: [],
    starship_class: "landing craft",
    url: "https://swapi.dev/api/starships/5/",
  },
  {
    MGLT: "10",
    cargo_capacity: "110",
    consumables: "1 week",
    cost_in_credits: "8000",
    created: "2014-12-15T12:31:42.547000Z",
    crew: "2",
    edited: "2014-12-20T21:23:49.897000Z",
    hyperdrive_rating: "1.0",
    length: "20",
    manufacturer: "Sienar Fleet Systems",
    max_atmosphering_speed: "1050",
    model: "TIE/LN starfighter",
    name: "TIE/LN starfighter",
    passengers: "0",
    films: [1], // Example film IDs
    pilots: [],
    starship_class: "Starfighter",
    url: "https://swapi.dev/api/starships/6/",
  },
];
