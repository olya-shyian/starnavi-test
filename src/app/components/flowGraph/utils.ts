import { FormattedFilms } from "@/app/interfaces/IFilms";
import { Edge, Node, Position } from "reactflow";
import dagre from "dagre";
import { EDGE_TYPE, NODE_HEIGHT, NODE_WIDTH, POSITION } from "./consts";
import { DirectionEnum } from "@/app/enums/DirectionEnum";

// Function to generate nodes for the graph
export const generateNodes = (
  heroeName: string,
  heroeInfo: FormattedFilms[]
) => {
  const nodes: Node[] = [
    {
      id: heroeName,
      data: { label: heroeName },
      style: { background: "#ff8c94", color: "#fff", borderColor: "#fff" },
      position: POSITION,
    }, // First node with the hero's name
  ];

  // Add nodes for each film in which the heroe appeared
  heroeInfo.forEach((film) => {
    const { title, starships } = film;

    nodes.push({
      id: title,
      data: { label: title },
      style: { background: "#3b5a9d", color: "#fff", borderColor: "#fff" },
      position: POSITION,
    });

    // Add nodes for each starship from the films
    starships &&
      starships.forEach((starship) => {
        const { name } = starship;
        nodes.push({
          id: name,
          data: { label: name },
          style: { background: "#35375a", color: "#fff", borderColor: "#fff" },
          position: POSITION,
        });
      });
  });

  return nodes;
};

// Function to generate edges for the graph
export const generateEdges = (
  heroeName: string,
  heroeInfo: FormattedFilms[]
) => {
  const edges: Edge[] = [];

  // Add connections from hero to each film
  heroeInfo.forEach((film, filmIndex) => {
    const { title, starships } = film;

    edges.push({
      id: `e${filmIndex + 1}-${heroeName}`,
      source: heroeName,
      target: title,
      type: EDGE_TYPE,
      animated: true,
    });

    // Add connections from each film to each starship
    starships &&
      starships.forEach((starship, starshipIndex) => {
        const { name } = starship;

        edges.push({
          id: `e${filmIndex + 1}-${starshipIndex + 1}-${title}`,
          source: title,
          target: name,
          type: EDGE_TYPE,
          animated: true,
        });
      });
  });

  return edges;
};

// Function to layout the elements using dagre (from right to left or from top to bottom)
export const getLayoutedElements = (
  nodes: Node[],
  edges: Edge[],
  direction = DirectionEnum.TopToBottom
) => {
  // Create a new dagre graph in order to set the direction for the layout of nodes in the graph.
  const dagreGraph = new dagre.graphlib.Graph();
  dagreGraph.setDefaultEdgeLabel(() => ({}));

  const isHorizontal = direction === DirectionEnum.LeftToRight;
  dagreGraph.setGraph({ rankdir: direction });

  nodes.forEach((node) => {
    dagreGraph.setNode(node.id, { width: NODE_WIDTH, height: NODE_HEIGHT });
  });

  edges.forEach((edge) => {
    dagreGraph.setEdge(edge.source, edge.target);
  });

  dagre.layout(dagreGraph);

  nodes.forEach((node) => {
    const nodeWithPosition = dagreGraph.node(node.id);

    node.targetPosition = isHorizontal ? Position.Left : Position.Top;
    node.sourcePosition = isHorizontal ? Position.Right : Position.Bottom;

    // We are shifting the dagre node position (anchor=center center) to the top left
    // so it matches the React Flow node anchor point (top left).
    node.position = {
      x: nodeWithPosition.x - NODE_WIDTH / 2,
      y: nodeWithPosition.y - NODE_HEIGHT / 2,
    };

    return node;
  });

  return { nodes, edges };
};
