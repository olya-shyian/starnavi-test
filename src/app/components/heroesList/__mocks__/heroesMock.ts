import { IHeroe } from "@/app/interfaces/IHeroe";

export const HeroesMock: IHeroe[] = [
  {
    birth_year: "19BBY",
    eye_color: "Blue",
    films: [1, 2, 3, 4, 5, 6],
    gender: "Male",
    hair_color: "Blond",
    height: "172",
    homeworld: 1,
    mass: "77",
    name: "Luke Skywalker",
    skin_color: "Fair",
    created: "2022-01-01T00:00:00.000Z",
    edited: "2022-01-01T00:00:00.000Z",
    species: [1],
    starships: [1, 2, 3],
    url: "https://swapi.dev/api/people/1/",
    vehicles: [1, 2],
  },
  {
    birth_year: "112BBY",
    eye_color: "Yellow",
    films: [1, 2, 3],
    gender: "Female",
    hair_color: "Brown",
    height: "150",
    homeworld: 2,
    mass: "49",
    name: "Leia Organa",
    skin_color: "Light",
    created: "2022-01-01T00:00:00.000Z",
    edited: "2022-01-01T00:00:00.000Z",
    species: [1],
    starships: [3, 4],
    url: "https://swapi.dev/api/people/2/",
    vehicles: [3, 4],
  },
];
