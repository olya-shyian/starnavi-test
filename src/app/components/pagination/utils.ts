import { NEXT_GROUP, PAGINATION_STEP, PREVIOUS_GROUP } from "./consts";

export const calculateTargetPageForDots = (
  page: number,
  currentPage: number
) => {
  let targetPage = page;

  if (page === NEXT_GROUP) {
    targetPage = currentPage + PAGINATION_STEP;
  } else if (page === PREVIOUS_GROUP) {
    targetPage = currentPage - PAGINATION_STEP;
  }

  return targetPage;
};
