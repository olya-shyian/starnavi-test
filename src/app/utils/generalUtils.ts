import { ReadonlyURLSearchParams } from "next/navigation";

// Extracts the ID from a given URL string

export const getIdFromUrl = (url: string): string | null => {
  const match = url.match(/(\d+)\/?$/);

  return match ? match[1] : null;
};

// Get a new searchParams string by merging the current
// searchParams with a provided key/value pair
export const createQueryString = (
  searchParams: ReadonlyURLSearchParams,
  name: string,
  value: string,
  paramsToRemove?: string[]
) => {
  const params = new URLSearchParams(searchParams.toString());
  params.set(name, value);

  if (paramsToRemove && !!paramsToRemove.length) {
    paramsToRemove.forEach((param) => {
      params.delete(param);
    });
  }

  return params.toString();
};
