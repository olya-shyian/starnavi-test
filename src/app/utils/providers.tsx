"use client";

import { ChakraProvider } from "@chakra-ui/react";

// A wrapper component that provides the Chakra UI theme and global styles
// to its children components. Other providers can be added here.
export const Providers = ({ children }: { children: React.ReactNode }) => (
  <ChakraProvider>{children}</ChakraProvider>
);
