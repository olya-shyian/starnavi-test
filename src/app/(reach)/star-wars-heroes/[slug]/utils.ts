import { FormattedFilms } from "@/app/interfaces/IFilms";
import { IStarships } from "@/app/interfaces/IStarships";
import { getIdFromUrl } from "@/app/utils/generalUtils";

// Combines starship data with film data to create a new array
// with film details and corresponding starship information.
export const addStarshipsToFilms = (
  data: IStarships[],
  formattedFilms: FormattedFilms[]
) => {
  return formattedFilms.map((film) => {
    const { id } = film;

    const starships = data.reduce(
      (acc: { id: string; name: string }[], { films, url, name }) => {
        if (films && films.includes(Number(id))) {
          const idFromUrl = getIdFromUrl(url);

          if (idFromUrl) {
            acc.push({ id: idFromUrl, name });
          }
        }

        return acc;
      },
      []
    );

    return { ...film, starships };
  });
};

// Get comma-separated ids from array of objects
export const getIdsAsString = (array: { id: string }[]): string => {
  const ids = array.map(({ id }: { id: string }) => id).join(",");

  return ids;
};
