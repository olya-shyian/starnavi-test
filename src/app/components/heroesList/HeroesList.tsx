import React from "react";
import { Box, Flex, Text, Divider } from "@chakra-ui/react";
import { IHeroe } from "@/app/interfaces/IHeroe";
import Link from "next/link";
import { createQueryString, getIdFromUrl } from "@/app/utils/generalUtils";
import { useSearchParams } from "next/navigation";

interface HeroesListProps {
  people: IHeroe[];
}

const HeroesList = ({ people }: HeroesListProps) => {
  const searchParams = useSearchParams();

  return (
    <Box m={10} p={3} height={430} borderWidth={2} borderRadius={6}>
      {people.map((data, index) => {
        const { name, url } = data;

        const heroesId = getIdFromUrl(url);
        const query = createQueryString(searchParams, "name", name);

        return (
          <React.Fragment key={name}>
            {index !== 0 && <Divider />}
            <Link href={`/star-wars-heroes/${heroesId}?` + query}>
              <Flex
                p={2}
                align="center"
                justify="center"
                _hover={{ fontWeight: "semibold" }}
              >
                <Text data-testid="hero-name">{name}</Text>
              </Flex>
            </Link>
          </React.Fragment>
        );
      })}
    </Box>
  );
};

export default HeroesList;
