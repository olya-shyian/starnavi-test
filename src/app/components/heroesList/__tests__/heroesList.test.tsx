import { render } from "@testing-library/react";
import HeroesList from "../HeroesList";
import { HeroesMock } from "../__mocks__/heroesMock";
import { createQueryString } from "@/app/utils/generalUtils";

// Mock useSearchParams
jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(() => new URLSearchParams("")),
}));

describe("HeroesList component", () => {
  it("renders a list of heroes", () => {
    const { asFragment } = render(<HeroesList people={HeroesMock} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
