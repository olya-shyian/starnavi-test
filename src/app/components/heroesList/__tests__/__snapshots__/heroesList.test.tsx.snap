// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesList component renders a list of heroes 1`] = `
<DocumentFragment>
  <div
    class="css-1m5dnc2"
  >
    <a
      href="/star-wars-heroes/1?name=Luke+Skywalker"
    >
      <div
        class="css-14ch23z"
      >
        <p
          class="chakra-text css-0"
          data-testid="hero-name"
        >
          Luke Skywalker
        </p>
      </div>
    </a>
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-1upb9tn"
    />
    <a
      href="/star-wars-heroes/2?name=Leia+Organa"
    >
      <div
        class="css-14ch23z"
      >
        <p
          class="chakra-text css-0"
          data-testid="hero-name"
        >
          Leia Organa
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
